File: 01-intro/03-basic-combinatorial-algorithms/02-variations/variations-iterative.ex
defmodule VariationLinar do
  def pow(base, p), do: :math.pow(base, p) |> round()

  def compute_slot(index, slot_index, elements_count) do
    index |> div(pow(elements_count, slot_index)) |> rem(elements_count)
  end

  def compute(elements, slots) do
    elements_count = Enum.count(elements)
    indexes = 0..(pow(elements_count, slots) - 1)

    for index <- indexes do
      slot_range = (slots - 1)..0 |> Enum.to_list()

      for slot_index <- slot_range do
        slot = compute_slot(index, slot_index, elements_count)
        Enum.at(elements, slot)
      end
    end
  end
end

IO.inspect(VariationLinar.compute(["a", "b", "c", "d"], 3))

File: 01-intro/03-basic-combinatorial-algorithms/02-variations/variations-iterative.py
class Variation(object):
    elements = ["a", "b",]
    element_count = 2
    slot_count = 5

    def compute_slot(self, index, slot_index):
        return (index // (self.element_count ** slot_index)) % \
               self.element_count

    def compute(self):
        for index in range(0, self.element_count ** self.slot_count):
            slot_range = range(self.slot_count - 1, -1, -1)
            element_indexes = [self.compute_slot(index, slot_index) for \
                    slot_index in \
                    slot_range]

            print([self.elements[index] for index in element_indexes])

if __name__ == "__main__":
    Variation().compute()